keps/sig-auth/4872-harden-kubelet-cert-validation/README.md

```diff
@@ -0,0 +1,389 @@
+
+## Summary
+
+This proposal aims to enhance the security of the Kube API server by validating the Common Name (CN) of the kubelet's serving certificate to ensure it matches the expected node name.
+This validation prevents a compromised node that has obtained a certificate for an IP address it does not own from using it to impersonate another node.
+
+## Motivation
+
+In cloud environments, IPs can change rapidly due to the ephemeral nature of the infrastructure.
+If IPs or machines rotate faster than the expiration frequency of kubelet serving certificates, a certificate issued to an old node could be used to respond to requests aimed at a new node, provided they share an IP.
+
+In addition, in on-premises environments, verifying that the IP addresses in a Certificate Signing Request (CSR) are owned by the requesting node can be challenging due to the lack of a reliable source of truth for IP ownership.
+Even when such a source exists, integration can be complex, leading to unsafe practices like auto-approval of CSRs without a strong guarantee of IP ownership.
+This vulnerability can be exploited through ARP poisoning or other routing attacks, allowing a rogue node to obtain a certificate for an IP it does not own and reroute traffic to itself.
+
+When the Kube API server connects to a kubelet, it verifies that the serving certificate is signed by a trusted CA and that the IP or hostname it’s connecting to is included in the certificate's SANs.
+If a rogue node obtained a certificate for an IP it does not own and reroute traffic to itself, it would be able to impersonate a Node that reports that IP.
+
+### Impact of node impersonation
+
+Provided an actor with control of a node can impersonate another node, the impact would be:
+
+* Break confidentiality of the requests sent by the Kube-API server to the kubelet (e.g kubectl exec/logs).These are usually user-driven requests. That gives the threat actor the possibility of producing incorrect or mis-leading feedback. In the exec case, it could allow a threat actor to issue prompts for credentials. In addition, the exec commands might contain user secrets.  
+* Break confidentiality of credentials if the client uses token based authentication. This is probably more common for non Kube-API server clients, given mTLS is common for Kube-API server to kubelet communication.
+
+### Goals
+
+* Ensure the Kube API server validates that the node’s serving certificate's CN matches the expected node name.  
+* Prevent rogue nodes from using certificates issued for IPs they do not own.
+
+### Non-Goals
+
+* This proposal does not address certificate validation for clients other than the Kube API server, such as metrics scrapers. However, we'll consider an implementation in client-go that could be used by those other clients.
+
+## Proposal
+
+We propose that the Kube API server is modified to validate the Common Name (CN) of the kubelet's serving certificate is equal to `system:node:<nodename>`.
```

Comment:
do you have a proof-of-concept of the wiring / wrapping for this validation? I poked around at the kube-apiserver → node network paths and didn't see an obvious spot to wire something at the connection level that had visibility to the node name and the serving certificate

Reply:
Hmm... thanks for the pointer... that's about as twisty as what I was afraid my exploration was going to turn into :-/

I think that approach is going to leak client transports in kube-apiserver... currently, there's a single transport for all kube-apiserver → kubelet connections. This approach looks like it will make a new transport per node, which gets held forever in the transport tlsCache ... so as nodes cycle in/out of a cluster, the kube-apiserver client cache will just grow unbounded :-/

we need to make sure we have a plausible way to hook in this validation that won't leak in that way

(cc @aojea @enj for client cache fun :-/)

Reply:
Indeed, that's why i was on the fence about the caching

The two alternatives I can see are to not cache these transports at all (either by making transports with custom cert validation uncachable or by allowing to mark a transport as uncachable and doing that explicitly for the ones we create) or to add a cache invalidation mechanism. Leaning towards the first one. The drawback is the overhead of creating a transport on the fly per request.

Reply:
@liggitt would you like to see a poc for either or those 2 options?
Or do you consider them both insufficient and I should look for something else in order to move forward with the kep?

Reply:
@g-gaston I am working on a POC of the client-go changes that should allow us to keep the TLS cache while adding the extra CN checks.  Will share soon.  cc @micahhausler

Reply:
@g-gaston @micahhausler @liggitt PTAL at https://github.com/kubernetes/kubernetes/pull/131337

Reply:
talked through possible approaches with @enj (notes in https://docs.google.com/document/d/1RqhAkGov_coHsB3lbAo-qfQl1MOfYvgpPUjiGMJ_3PY/edit?tab=t.0, shared with dev@kubernetes.io and sig-auth mailing list)

the approach in https://github.com/kubernetes/kubernetes/pull/131450 looks the most likely